src/js/common/lib/i18n.js

```diff
@@ -0,0 +1,188 @@
+ *
+ */
+gpii.handlebars.i18n.getAllLocalesFromHeader = function (header) {
+    var acceptedLanguages = [];
+    if (header === undefined) {
+        acceptedLanguages.push("*");
+    }
+    else if (typeof header === "string") {
+        var localeSegments = header.trim().split(/ *, */);
+        var segmentsAsObjects = [];
+        fluid.each(localeSegments, function (localeSegment) {
+            var subSegments = localeSegment.split(";q=");
+            var segmentDef = {
+                locale: subSegments[0],
+                q: subSegments.length > 1 ? subSegments[1] : 1
+            };
+            segmentsAsObjects.push(segmentDef);
+        });
+        segmentsAsObjects.sort(gpii.handlebars.i18n.sortByQScore);
+        segmentsAsObjects.map(function (entry) {
+            acceptedLanguages.push(entry.locale);
+        });
+    }
+    return acceptedLanguages.map(function (entry) { return entry.toLowerCase().replace("-", "_");});
+};
+
+/**
+ *
+ * Sort by the "q score" included with complex "Accept-Language" headers.
+ *
+ * @param {Object} a - A single weighted language entry.
+ * @param {Object} b - Another weighted language to compare to `a`.
+ * @return {Number} - -1 if `a` should come before `b`, 1 if `b` should come before `a`, 0 if their weights are equal.
+ *
+ */
+gpii.handlebars.i18n.sortByQScore = function (a, b) {
+    if (a.q === b.q) {
+        return 0;
+    }
+    else if (a.q > b.q) {
+        return -1;
+    }
+    else {
+        return 1;
+    }
+};
+
+/**
+ *
+ * Derive the effective message bundle from the value of the `Accept-Language` header.
+ *
+ * @param {String} header - The `Accept-Language` header provided by the browser.  See: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Language
+ * @param {Object} messageBundles - The "bundle of message bundles" representing messages for all available languages.
+ * @param {String} defaultLocale - The default locale to use if no language is specified (or if a wildcard is specified).
+ * @return {Object} A message bundle containing messages for the most appropriate combination of locale, language, and defaults.
+ *
+ */
+gpii.handlebars.i18n.deriveMessageBundleFromHeader = function (header, messageBundles, defaultLocale) {
+    var localesAndLanguages = gpii.handlebars.i18n.languagesAndLocalesFromMessageBundle(messageBundles, defaultLocale);
+    var allLocalesFromHeader = gpii.handlebars.i18n.getAllLocalesFromHeader(header);
+
+    var inheritedMatchingLanguage = fluid.find(allLocalesFromHeader, function (candidateLocale) {
+        var language = gpii.handlebars.i18n.languageFromLocale(candidateLocale);
+        if (language && localesAndLanguages.indexOf(language) !== -1) {
+            return language;
+        }
+    });
+
+    var directMatchingLocaleOrLanguage = fluid.find(allLocalesFromHeader, function (candidateLocale) {
+        if (candidateLocale === "*") {
+            return defaultLocale;
```

Comment:
Untested branch

Reply:
I reworked the tests to better exercise the core functions and rewrote based on oddities revealed.  Previously the failover from locale -> language -> defaults was not completely consistent, it is now.